### PR TITLE
mrsdyna: New driver for Mrs. Dynamite

### PR DESCRIPTION
**Pull request by palazzol**
I also added a lot of hardware notes for the Universal 8203 boardset.  
I expect to add one more revision to document the rest of it.

### CODE REVIEW

File: src/mame/drivers/ladybug.cpp [hunk 1]
```diff
@@ -135,42 +143,82 @@ void dorodon_state::decrypted_opcodes_map(address_map &map)
...
-	map(0x9003, 0x9003).portr("DSW1");
-	map(0xd000, 0xd7ff).w("video", FUNC(ladybug_video_device::bg_w));
-	map(0xe000, 0xe000).nopw();  //unknown 0x10 when in attract, 0x20 when coined/playing
+	// LS138 @ J4
+	map(0x0000, 0x1fff).rom();	// 2764 at R4
+	map(0x2000, 0x3fff).rom();	// 2764 at N4
+	map(0x4000, 0x5fff).rom();	// 2764 at M4
+	// LS138 @ J4 and LS139 @ H4
+	map(0x6000, 0x67ff).ram();  // 6116 @ K3, also clk on PAL K2 (16R6, U001)
+	map(0x6800, 0x6fff).ram();  // 6116 @ M3
```
Comment (cuavas) [inline src/mame/drivers/ladybug.cpp:154]: Don’t split up handlers for contiguous ROM/RAM – it’s inefficient.

File: src/mame/drivers/ladybug.cpp [hunk 2]
```diff
@@ -135,42 +143,82 @@ void dorodon_state::decrypted_opcodes_map(address_map &map)
...
+	map(0x7000, 0x73ff).w("video", FUNC(ladybug_video_device::spr_w)); // pin 29 on ribbon 
+	//map(0x77ff, 0x7fff);	// LS139 @ H4 pin7 is NC
+
+	// LS138 @ J4, Pin11 (0x8000-0x9fff) and
+	// LS138 @ N3 (bottom 3 bits)
+	// (all of these are read/write)
+	map(0x8005, 0x8005).mirror(0x1ff8).r(FUNC(mrsdyna_state::mrsdyna_protection_r));  // OE on PAL @ K2 (16R6, U001) (100x xxxx xxxx x101)
+	map(0x8006, 0x8006).mirror(0x1ff8).writeonly().share("sound_low");  // LS374 @ P6
+	map(0x8007, 0x8007).mirror(0x1ff8).writeonly().share("sound_high"); // LS374 @ R6
```
Comment (cuavas) [inline src/mame/drivers/ladybug.cpp:163]: Would it be better to use `latch8_device` for these so you get an automatic synch barrier on write?

File: src/mame/drivers/ladybug.cpp [hunk 3]
```diff
@@ -790,7 +938,6 @@ void sraider_state::sraider(machine_config &config)
 	SN76489(config, "sn5", 4000000).add_route(ALL_OUTPUTS, "mono", 1.0);
 }
 
-
```
Comment (cuavas) [inline src/mame/drivers/ladybug.cpp:793]: Please don’t go removing extra whitespace separating sections in files.

File: src/mame/drivers/ladybug.cpp [hunk 4]
```diff
@@ -994,7 +1169,6 @@ ROM_START( sraider )
 	ROM_LOAD( "srpr10-3.c1",  0x0040, 0x0020, CRC(27fa3a50) SHA1(7cf59b7a37c156640d6ea91554d1c4276c1780e0) ) /* ?? */
 ROM_END
 
-
```
Comment (cuavas) [inline src/mame/drivers/ladybug.cpp:997]: This was also separating sections with a larger vertical gap.

File: src/mame/mame.lst [hunk 5]
```diff
@@ -19557,7 +19557,8 @@ ladybug                         // 8110 (c) 1981
 ladybugb                        // bootleg
 ladybugb2                       // bootleg
 snapjack                        // ???? (c)
-sraider                         // ???? (c) 1982 Universal
+mrsdyna                         // 8203 (c) 1982 Universal
```
Comment (cuavas) [inline src/mame/mame.lst:19560]: Please put this in alphabetical order within the section.

File: src/mame/drivers/ladybug.cpp (hunk 1 above)
Comment (DavidHaywood) [reply src/mame/drivers/ladybug.cpp:154]: in cases like this it seems better documentation this way, which is most important?

Comment (cuavas) [reply src/mame/drivers/ladybug.cpp:154]: You can still put which chip covers which range in a comment.

**Comment by palazzol** [comment]
I resolved the issues, and used latch8_device for the sound latches.  Didn't know about that one - seems like a good idea to me too.

File: src/mame/includes/ladybug.h [hunk 6]
```diff
@@ -13,6 +13,7 @@
 #include "video/ladybug.h"
 #include "emupal.h"
 #include "tilemap.h"
+#include "machine/latch8.h"
```
Comment (cuavas) [inline src/mame/includes/ladybug.h:16]: Please put this above the headers from src/emu (organise modules from most dependent to least dependent, to catch headers that don’t include dependencies faster).

File: src/mame/includes/ladybug.h [hunk 7]
```diff
@@ -84,52 +85,80 @@ class dorodon_state : public ladybug_state
...
+	void mrsdyna_cpu2_map(address_map &map);
 
-private:
-	required_shared_ptr<uint8_t> m_grid_data;
 	required_device<palette_device> m_palette;
 	required_device<gfxdecode_device> m_gfxdecode;
 	required_device<ladybug_video_device> m_video;
-	required_device<zerohour_stars_device> m_stars;
+	required_device<latch8_device> m_soundlatch_low;
+	required_device<latch8_device> m_soundlatch_high;
```
Comment (cuavas) [inline src/mame/includes/ladybug.h:126]: There’s no point having these here if you don’t actually use them in the machine configuration and address map.  Either use the object finders in place of explicit tags in the address maps and machine configuration, or don’t add these object finder members (if you choose the latter, you can also punt the `#include "machine/latch8.h"` over to the .cpp file).

Comment (palazzol) [reply src/mame/includes/ladybug.h]: I went over this in detail with @smf- and I understand how it works now.  I chose the latter.

**Pull request merged**